Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +376,120 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
+                    return
+                }
+
+                /// http_parser_execute(...) should always consume the whole buffer.
+                assert(result == bufferSlice.readableBytes)
+
+                // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+                self.cumulationBuffer!.moveReaderIndex(forwardBy: result)
+
+                if bufferSlice.readableBytes == self.cumulationBuffer!.readableBytes + result {
+                    // There was no change to the readable bytes of the cumulationBuffer by a re-entrant call of channelRead. Its safe to break the loop.
+                    break
+                }
+            } while true
+
+            switch state.dataAwaitingState {
+            case .body, .messageBegin:
+                // Ensure there is no risk to invalidate any indices that we depend on.
+                assert(state.currentNameIndex == nil)
+                assert(state.currentHeaders.isEmpty)
+                assert(state.slice == nil)
+
+                // Only in the body and messageBegin states its safe to discard read bytes as otherwise we may risk to invalidate the stored indices for the
+                // Headers and slice.
+                if self.shouldReclaimBytes(buffer: self.cumulationBuffer!) {
+                    self.cumulationBuffer!.discardReadBytes()
+                }
+            default:
+                // TODO: We may be able to also discard some read bytes in other states if we are very careful about it.
+                //       What we would need to do is basically keep track of the start of a message and discard everything that we already forwarded through the pipeline.
+                break
             }
-            return result
+        } catch {
+            self.cumulationBuffer = nil
+            ctx.fireErrorCaught(error)
+            ctx.close(promise: nil)
         }
+    }
 
-        if let slice = state.slice {
-            // If we have a slice, we need to preserve all of these bytes. To do that, we move the
-            // reader index to where the slice wants it, and then record how many readable bytes that leaves
-            // us with. Then, invalidate the reader index, as it's not stable across calls to decode()
-            // *anyway*, so we want to make sure we can see the bad value in debug errors.
-            buffer.moveReaderIndex(to: slice.readerIndex)
-            state.readerIndexAdjustment = buffer.readableBytes
-            state.slice = (-1, slice.length)
-            return .needMoreData
-        } else {
-            buffer.moveReaderIndex(forwardBy: result)
-            state.readerIndexAdjustment = 0
-            return .continue
-        }
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        return DecodingState.needMoreData
```

Comment:
Can you add a code comment here indicating that we don't actually expect this function to be called?